### PR TITLE
[ Bugfix ] Enabling Loading Models With Fused QKV/MLP on Disk with FP8

### LINKED ISSUE #5915: [Bug]: FP8 checkpoints with fused linear modules fail to load scales correctly
### Your current environment

```text
The output of `python collect_env.py`
```

### 🐛 Describe the bug

Description:
When loading FP8 quantized models with merged linear modules (e.g., Phi-3 with merged qkv_proj and up_gate_proj), the scales for each shard are not handled correctly. This occurs because the vLLM FP8 config assumes separate scales for each shard, but merged layers have a single scale.

Steps to Reproduce:
1. Attempt to load an FP8 quantized Phi-3 model (e.g., https://huggingface.co/nm-testing/Phi-3-mini-128k-instruct-FP8)
2. Observe error due to shape mismatch:
   ```
   param_data.shape=torch.Size([2]) loaded_weight.shape=torch.Size([])
   param_data.shape=torch.Size([3]) loaded_weight.shape=torch.Size([])
   ```

Expected Behavior:
Scales should be correctly loaded for merged linear modules in FP8 checkpoints.

Proposed Fix:
Modify `process_weights_after_loading` in MergedColumnParallelLinear and QKVParallelLinear to repeat the merged scale during weight loading.

Temporary Workaround:
Apply the following patch in `vllm/model_executor/layers/linear.py`:
```python
- assert param_data.shape == loaded_weight.shape
- param_data.copy_(loaded_weight)
+ temp = loaded_weight.repeat(param_data.shape)
+ assert param_data.shape == temp.shape
+ param_data.copy_(temp)
```

cc @robertgshaw2-neuralmagic @comaniac

### PR DESCRIPTION
**Pull request by robertgshaw2-redhat**
SUMMARY:
* some models have fused q,k,v and fused mlp on disk (e.g. Phi3)
* our fp8 implementation assumes there is a separate key for each q,k,v during weight loading. Because our kernels require a single scale (`torch.scaled_mm`), we load 3 scales for `QKVColumnLinear`. Then, after the weights have been loaded, we select the biggest scale of the group, dequantize each logical weight individually and re-quantize with the biggest scale
* this PR fixes the logic by (a) initializing all the scales to min value, (b) skipping the dequant-->quant logic if we detect the model has shared qkv on disk

NOTE:
* I will admit, this is somewhat of a hacky solution where we are special casing a lot. To handle this cleanly, I think we will need to refactor fp8. @comaniac WDYT

FIX #5915

**BEFORE SUBMITTING, PLEASE READ THE CHECKLIST BELOW AND FILL IN THE DESCRIPTION ABOVE**

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

File: vllm/model_executor/layers/quantization/fp8.py [hunk 1]
```diff
@@ -98,6 +98,7 @@ class Fp8LinearMethod(LinearMethodBase):
     """
 
     def __init__(self, quant_config: Fp8Config):
+        self.qkv_mlp_fused_on_disk = False
```
Comment (mgoin) [inline vllm/model_executor/layers/quantization/fp8.py:101]: Maybe `fused_module_in_checkpoint` would be a bit more clear and generalized

File: vllm/model_executor/layers/quantization/fp8.py [hunk 2]
```diff
@@ -227,10 +235,6 @@ def process_weights_after_loading(self, layer: Module) -> None:
             if self.quant_config.activation_scheme == "dynamic":
                 layer.input_scale = None
             elif self.quant_config.activation_scheme == "static":
-                if not all_close_1d(layer.input_scale):
-                    raise ValueError(
-                        "All the input_scales for the logical weights of a "
-                        f"layer must be equal. But got {layer.input_scale}")
```
Comment (mgoin) [inline vllm/model_executor/layers/quantization/fp8.py:233]: Why is this check not needed anymore?

File: vllm/model_executor/layers/quantization/fp8.py [hunk 3]
```diff
@@ -111,6 +112,7 @@ def _create_scale_param(
         scale = Parameter(torch.empty(len(output_partition_sizes),
                                       dtype=torch.float32),
                           requires_grad=False)
+        scale[:] = torch.finfo(torch.float8_e4m3fn).min
```
Comment (comaniac) [inline vllm/model_executor/layers/quantization/fp8.py:115]: What's the purpose of this?

**Review by comaniac — APPROVED** [review]
Overall LGTM, but yeah these logics become messy and we should refactor along with linear layers. Hopefully the weight loading refactoring could help with this.

One side note: I also have a requirement that loads a sharded state dicts directly and bypass the normal weight loading process. In this case the weight shapes are mismatching because we initialized the weights in the original shape and transpose after weight loading. We should avoid this and make sure the initialized weights are already in the final shape.

**Review by mgoin — APPROVED** [review]
I validated this works with the Phi 3 mini FP8 models we have, thanks!

**Comment by robertgshaw2-neuralmagic** [comment]
> Overall LGTM, but yeah these logics become messy and we should refactor along with linear layers. Hopefully the weight loading refactoring could help with this.
> 
> One side note: I also have a requirement that loads a sharded state dicts directly and bypass the normal weight loading process. In this case the weight shapes are mismatching because we initialized the weights in the original shape and transpose after weight loading. We should avoid this and make sure the initialized weights are already in the final shape.

yeah, dipika's work should help with this. Though Im not quite sure how we can avoid the 3 vs 1 scales initialized issue since we don't know that the model is serialized fused until after we peek into the safetensors

File: vllm/model_executor/layers/quantization/fp8.py (hunk 2 above)
Comment (robertgshaw2-neuralmagic) [reply vllm/model_executor/layers/quantization/fp8.py:233]: * We always initialize 3 scales in QKVColumnLinear
* In the case of fused QKV on disk, there is only 1 scale on disk for QKV
* In this PR, I load the scale from disk into the first index of the scales in QKVColumnLinear (and the other 2 are set to -inf)
* So it is no longer true that these all need to be "close"

**Comment by robertgshaw2-neuralmagic** [comment]
Thanks for the quick review!

**Comment by DarkLight1337** [comment]
To speed up the CI queue for #5905, I've cancelled the distributed tests for the latest CI run in this PR since they won't pass anyway until #5905 has been merged. Please merge `main` into your branch after that happens so that the CI can pass once again.
